pkg/cgroups/cgroupsv2/runtime.go

```diff
@@ -0,0 +1,147 @@
+package cgroupsv2
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"syscall"
+	"time"
+
+	"github.com/google/uuid"
+	jobv1 "github.com/kralicky/jobserver/pkg/apis/job/v1"
+	"github.com/kralicky/jobserver/pkg/cgroups"
+	"github.com/kralicky/jobserver/pkg/jobs"
+	"github.com/kralicky/jobserver/pkg/util"
+)
+
+const gracePeriod = 10 * time.Second
+
+type v2Runtime struct {
+	mgr *cgroupManager
+}
+
+func NewRuntime() (jobs.Runtime, error) {
+	mgr, err := newCgroupManager()
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup jobserver cgroup: %w", err)
+	}
+
+	return &v2Runtime{
+		mgr: mgr,
+	}, nil
+}
+
+// Execute implements jobs.Runtime.
+func (l *v2Runtime) Execute(ctx context.Context, spec *jobv1.JobSpec) (jobs.Process, error) {
+	cmdSpec := spec.GetCommand()
+
+	// generate a uuid for the process, but encode it in the raw hex format.
+	// this makes it slightly easier to use from the command line, since many
+	// terminals treat '-' as a word separator, making it difficult to copy
+	// the whole string using a double-click.
+	u := uuid.New()
+	id := hex.EncodeToString(u[:])
+
+	cmd := exec.CommandContext(ctx, cmdSpec.GetCommand(), cmdSpec.GetArgs()...)
+	cmd.Env = append(os.Environ(), cmdSpec.GetEnv()...)
+
+	streamBuf := util.NewStreamBuffer()
+	done := make(chan struct{})
+
+	cmd.Stdout = streamBuf
+	cmd.Stderr = streamBuf
+	cmd.Stdin = nil
+	cmd.Cancel = func() error {
+		lg := slog.With("id", id)
+		lg.Debug("context canceled; attempting graceful shutdown")
+		streamBuf.Close() // NB: leaving this open will cause cmd.Wait to hang
```

Comment:
Does this mean that we're going to miss some output in open streams?

Reply:
This is just referring to the fact that streamBuf indirectly holds open actual file descriptors (stdout/stderr pipes) that must be closed before cmd.Wait can return.

Reply:
I'll check if using WaitDelay can simplify some of this logic